File: data structures and algorithms/java/06_112.java
public class FibonacciVariations {
    // Without memoization - exponential time complexity
    public static long fibSlow(int n) {
        if (n <= 1) {
            return n;
        }
        return fibSlow(n-1) + fibSlow(n-2);
    }

    // With memoization using a HashMap
    public static long fibMemo(int n, Map<Integer, Long> memo) {
        if (memo == null) {
            memo = new HashMap<>();
        }
        
        if (memo.containsKey(n)) {
            return memo.get(n);
        }
        
        if (n <= 1) {
            return n;
        }
        
        memo.put(n, fibMemo(n-1, memo) + fibMemo(n-2, memo));
        return memo.get(n);
    }
    
    // Using memoization with a static cache
    private static Map<Integer, Long> cache = new HashMap<>();
    
    public static long fibCache(int n) {
        if (cache.containsKey(n)) {
            return cache.get(n);
        }
        
        if (n <= 1) {
            return n;
        }
        
        cache.put(n, fibCache(n-1) + fibCache(n-2));
        return cache.get(n);
    }
    
    public static void main(String[] args) {
        // This calculates quickly!
        System.out.println(fibCache(40));
    }
}

File: data structures and algorithms/cs/06_112.cs
using System;
using System.Collections.Generic;

public class FibonacciVariations {
    // Without memoization - exponential time complexity
    public static long FibSlow(int n) {
        if (n <= 1) {
            return n;
        }
        return FibSlow(n-1) + FibSlow(n-2);
    }

    // With memoization using a Dictionary
    public static long FibMemo(int n, Dictionary<int, long> memo = null) {
        if (memo == null) {
            memo = new Dictionary<int, long>();
        }
        
        if (memo.ContainsKey(n)) {
            return memo[n];
        }
        
        if (n <= 1) {
            return n;
        }
        
        memo[n] = FibMemo(n-1, memo) + FibMemo(n-2, memo);
        return memo[n];
    }
    
    // Using memoization with a static cache
    private static Dictionary<int, long> cache = new Dictionary<int, long>();
    
    public static long FibCache(int n) {
        if (cache.ContainsKey(n)) {
            return cache[n];
        }
        
        if (n <= 1) {
            return n;
        }
        
        cache[n] = FibCache(n-1) + FibCache(n-2);
        return cache[n];
    }
    
    public static void Main(string[] args) {
        // This calculates quickly!
        Console.WriteLine(FibCache(40));
    }
}